docs: add quickstart guide for k3s (#454)
submariner-io/submariner-website: src/content/getting-started/quickstart/k3s/_index.md

---
date: 2021-02-24T13:36:18+01:00
title: "K3s"
weight: 15
---

This quickstart guide covers the necessary steps to deploy two [K3s](https://k3s.io/) clusters
and deploy Submariner with Service Discovery to interconnect the two clusters. Note that this
guide focuses on Submariner deployment on clusters with non-overlapping Pod and Service CIDRs.
For connecting clusters with overlapping CIDRs, please refer to the
[Globalnet guide](../../architecture/globalnet/).

### Prerequisites

1. Prepare two nodes with
[K3s-supported OS](https://rancher.com/docs/k3s/latest/en/installation/installation-requirements/#operating-systems)
installed (These two nodes are referred to as node-a and node-b).
2. Choose the Pod CIDR and the Service CIDR for each K3s cluster.

    In this guide, we will use the following CIDRs:

    | Cluster   | Pod CIDR     | Service CIDR |
    |-----------|--------------|--------------|
    | cluster-a |10.44.0.0/16  |10.45.0.0/16  |
    | cluster-b |10.144.0.0/16 |10.145.0.0/16 |

### Deploy K3s Clusters

#### Deploy cluster-a on node-a

> In this step you will deploy **cluster-a** on node-a using the the specific CIDRs.
> To deploy K3s with the CIDRs, run the following commands on node-a.

Reviewer: ```suggestion
In this step you will deploy K3s on node-a using the the **cluster-a** CIDRs.
```

```bash
POD_CIDR=10.44.0.0/16
SERVICE_CIDR=10.45.0.0/16
curl -sfL https://get.k3s.io | INSTALL_K3S_EXEC="--cluster-cidr $POD_CIDR --service-cidr $SERVICE_CIDR" sh -s -
```

{{% notice tip %}}
For more information on interacting with K3s, please refer to the [k3s documentation](https://rancher.com/docs/k3s/latest/en/quick-start/).
{{% /notice %}}

> The kubeconfig file needs to be modified to set the Kubernetes API endpoint to the public IP of the node-a,

Reviewer: ```suggestion
The kubeconfig file needs to be modified to set the Kubernetes API endpoint to the public IP of node-a,
```

which is obtained from the first field yielded by `hostname -I`.

```bash
cp /etc/rancher/k3s/k3s.yaml kubeconfig.cluster-a
IP=$(hostname -I | awk '{print $1}')
sed -i 's/127.0.0.1/'$IP'/' kubeconfig.cluster-a
```

#### Deploy cluster-b on node-b

> In this step you will deploy **cluster-b** on node-b using the the specific CIDRs.
> To deploy K3s with the CIDRs, run the following commands on node-b.

Reviewer: ```suggestion
In this step you will deploy K3s on node-b using the the **cluster-b** CIDRs.
```

```bash
POD_CIDR=10.144.0.0/16
SERVICE_CIDR=10.145.0.0/16
curl -sfL https://get.k3s.io | INSTALL_K3S_EXEC="--cluster-cidr $POD_CIDR --service-cidr $SERVICE_CIDR" sh -s -
```

> The kubeconfig file needs to be modified to set the Kubernetes API endpoint to the public IP of the node-b,

Reviewer: ```suggestion
The kubeconfig file needs to be modified to set the Kubernetes API endpoint to the public IP of node-b,
```

which is obtained from the first field yielded by `hostname -I`.

```bash
cp /etc/rancher/k3s/k3s.yaml kubeconfig.cluster-b
IP=$(hostname -I | awk '{print $1}')
sed -i 's/127.0.0.1/'$IP'/' kubeconfig.cluster-b
```

Next, copy kubeconfig.cluster-b to node-a.

#### Install `subctl` on node-a

{{< subctl-install >}}

#### Use cluster-a as the Broker

On node-a, run:

```bash
subctl deploy-broker --kubeconfig kubeconfig.cluster-a --service-discovery
```

> #### Join cluster-a and cluster-b to the Broker
> 
> Join cluster-a to the Broker

Reviewer: I think it's simpler to just make this its own section - avoids repeating "Join cluster-a ...".

```suggestion
#### Join cluster-a to the Broker
```

```bash
subctl join --kubeconfig kubeconfig.cluster-a broker-info.subm --clusterid cluster-a --natt=false
```

> Join cluster-b to the Broker

Reviewer: ```suggestion
#### Join cluster-b to the Broker
```

---
date: 2021-02-24T13:36:18+01:00

> title: "k3s"

Reviewer 2: ```suggestion
title: "K3s"
```

It seems like it's generally capitalized in the project's docs? https://github.com/k3s-io/k3s

(applies elsewhere)

> weight: 10

Reviewer 2: Let's make this weight 15 (or 20 and bump the current 20/30 by 10) so it comes after the KIND sandbox/CI environment in the left nav bar list. I think that's our recommended "default path" for folks.

---

> This quickstart guide covers the necessary steps to deploy two [K3s](https://k3s.io/) cluslters

Reviewer: ```suggestion
This quickstart guide covers the necessary steps to deploy two [K3s](https://k3s.io/) clusters
```

and deploy Submariner with Service Discovery to interconnect the two clusters. Note that this
guide focuses on Submariner deployment on clusters with non-overlapping Pod and Service CIDRs.
For connecting clusters with overlapping CIDRs, please refer to the
[Globalnet guide](../../architecture/globalnet/).

### Prerequisites

1. Prepare two nodes with
[K3s-supported OS](https://rancher.com/docs/k3s/latest/en/installation/installation-requirements/#operating-systems)
installed (These two nodes are referred to as node-a and node-b).

> 2. Decide the Pod CIDR and the Service CIDR for each K3s cluster.

Reviewer: ```suggestion
2. Choose the Pod CIDR and the Service CIDR for each K3s cluster.
```

### Deploy K3s Clusters

#### Deploy cluster-a on node-a

> In this step you will deploy **cluster-a** on node-a using the the following IP ranges:

Reviewer: ```suggestion
In this step you will deploy **cluster-a** on node-a using the the following CIDR:
```

| Pod CIDR     | Service CIDR |
|--------------|--------------|

> |10.44.0.0/16  |10.45.0.0/16  |

Reviewer: In  the Prerequisites, we instruct the user to choose the CIDRs but here we're telling them to use specific CIDRs.

> To deploy K3s with the IP ranges, run below commands on node-a.

Reviewer: ```suggestion
To deploy K3s with the CIDRs, run the following commands on node-a.
```

```bash
POD_CIDR=10.44.0.0/16
SERVICE_CIDR=10.45.0.0/16
curl -sfL https://get.k3s.io | INSTALL_K3S_EXEC="--cluster-cidr $POD_CIDR --service-cidr $SERVICE_CIDR" sh -s -
```

{{% notice tip %}}
For more information on interacting with K3s, please refer to the [k3s documentation](https://rancher.com/docs/k3s/latest/en/quick-start/).
{{% /notice %}}

> The kubeconfig file needs to be modified like below to make k8s API endpoint point to public IP of the node-a.
> If the IP should be different from the first field of `hostname -I`, modify the value manually.

Reviewer: ```suggestion
The kubeconfig file needs to be modified to set the Kubernetes API endpoint point to the public IP the node-a,
which is obtained from the first field yielded by `hostname -I`.
```

```bash
cp /etc/rancher/k3s/k3s.yaml kubeconfig.cluster-a
IP=$(hostname -I | awk '{print $1}')
sed -i 's/127.0.0.1/'$IP'/' kubeconfig.cluster-a
```

#### Deploy cluster-b on node-b

> In this step you will deploy **cluster-b** on node-b using the the following IP ranges:

Reviewer: ```suggestion
In this step you will deploy **cluster-b** on node-b using the the following CIDRs:
```

| Pod CIDR     | Service CIDR |
|--------------|--------------|
|10.144.0.0/16 |10.145.0.0/16 |

> To deploy K3s with the IP ranges, run below commands on node-b.

Reviewer: ```suggestion
To deploy K3s with the CIDRs, run the following commands on node-b.
```

```bash
POD_CIDR=10.144.0.0/16
SERVICE_CIDR=10.145.0.0/16
curl -sfL https://get.k3s.io | INSTALL_K3S_EXEC="--cluster-cidr $POD_CIDR --service-cidr $SERVICE_CIDR" sh -s -
```

> The kubeconfig file needs to be modified like below to make k8s API endpoint point to public IP of the node-b.
> If the IP should be different from the first field of `hostname -I`, modify the value manually.

Reviewer: ```suggestion
The kubeconfig file needs to be modified to set the Kubernetes API endpoint point to the public IP the node-a,
which is obtained from the first field yielded by `hostname -I`.
```

```bash
cp /etc/rancher/k3s/k3s.yaml kubeconfig.cluster-b
IP=$(hostname -I | awk '{print $1}')
sed -i 's/127.0.0.1/'$IP'/' kubeconfig.cluster-b
```

> Then, copy kubeconfig.cluster-b to node-a.

Reviewer: ```suggestion
Next, copy kubeconfig.cluster-b to node-a.
```

#### Install `subctl` on node-a

{{< subctl-install >}}

> #### Use cluster-a as Broker

Reviewer: ```suggestion
#### Use cluster-a as the Broker
```

> On node-a, run the below command.

Reviewer: ```suggestion
On node-a, run:
```

```bash
subctl deploy-broker --kubeconfig kubeconfig.cluster-a --service-discovery
```

#### Join cluster-a and cluster-b to the Broker

On node-a, run the below commands.

##### Join node-a

```bash
subctl join --kubeconfig kubeconfig.cluster-a broker-info.subm --clusterid cluster-a --natt=false
```

##### Join node-b

```bash
subctl join --kubeconfig kubeconfig.cluster-b broker-info.subm --clusterid cluster-b --natt=false

> ```

Reviewer: I think we can reduce these sections.

#### Join cluster-a to the Broker

```bash
subctl join --kubeconfig kubeconfig.cluster-a broker-info.subm --clusterid cluster-a --natt=false
```

#### Join cluster-b to the Broker

```bash
subctl join --kubeconfig kubeconfig.cluster-b broker-info.subm --clusterid cluster-b --natt=false
```

> You now have a Submariner environment that you can experiment with.

Reviewer: I don't think we need this statement.

```suggestion
```
